src/atomicStore.ts

```diff
@@ -0,0 +1,278 @@
+    const desc = Object.getOwnPropertyDescriptor(initial, k);
+    if (
+      !desc ||
+      typeof desc.value !== 'function' || // Not an action
+      desc.get // Skip getters (derived state)
+    )
+      continue;
+
+    store[k] = createActionAtom(
+      k,
+      desc.value,
+      store,
+      baseAtoms,
+      initial,
+    ) as AtomicStore<State>[typeof k];
+  }
+
+  return store;
+}
+
+/**
+ * Creates a derived state atom that automatically updates when its dependencies change.
+ *
+ * @template T - The type of the atomic state definition object.
+ * @param key - The key of the derived state property.
+ * @param getFn - The getter function for the derived state.
+ * @param store - The atomic store.
+ * @param baseAtoms - Map of base atoms.
+ * @param initial - The initial atomic state definition.
+ * @returns An atom representing the derived state.
+ */
+function createDerivedAtom<T>(
+  key: keyof T,
+  getFn: () => any,
+  store: AtomicStore<T>,
+  baseAtoms: Map<keyof T, PrimitiveAtom<any>>,
+  initial: AtomicState<T>,
+): Atom<any> {
+  return atom((get) => {
+    const state = createStateGetter(get, store, baseAtoms, initial);
+    return getFn.call(state);
+  });
+}
+
+/**
+ * Creates an action atom that can update multiple base state values atomically.
+ *
+ * @template T - The type of the atomic state definition object.
+ * @param key - The key of the action property.
+ * @param actionFn - The function representing the action.
+ * @param store - The atomic store.
+ * @param baseAtoms - Map of base atoms.
+ * @param initial - The initial atomic state definition.
+ * @returns A writable atom representing the action.
+ */
+function createActionAtom<T>(
+  key: keyof T,
+  actionFn: Function,
+  store: AtomicStore<T>,
+  baseAtoms: Map<keyof T, PrimitiveAtom<any>>,
+  initial: AtomicState<T>,
+): WritableAtom<void, any[], void> {
+  type Args = T[typeof key] extends (...args: infer P) => any ? P : never;
+
+  return atom(null, (get, set, ...args: Args) => {
+    const state = createStateGetter(get, store, baseAtoms, initial);
+    const result = actionFn.apply(state, args);
+    if (result) {
+      for (const [k, v] of Object.entries(result)) {
+        if (baseAtoms.has(k as keyof T)) set(baseAtoms.get(k as keyof T)!, v);
+      }
+    }
+  }) as unknown as WritableAtom<void, Args, void>;
+}
+
+/**
+ * Creates an object with getters that access the latest atom values.
+ * Used to provide the `this` context in derived state and actions.
+ *
+ * @template T - The type of the atomic state definition object.
+ * @param get - The 'get' function provided by Jotai atoms.
+ * @param store - The atomic store.
+ * @param baseAtoms - Map of base atoms.
+ * @param initial - The initial atomic state definition.
+ * @returns An object with getters for each property.
+ */
+function createStateGetter<T>(
+  get: any,
+  store: AtomicStore<T>,
+  baseAtoms: Map<keyof T, PrimitiveAtom<any>>,
+  initial: AtomicState<T>,
```

Comment:
I don't think we need to pass this. Just loop over baseAtoms and store?

Reply:
I just inlined the helper methods - they were quite tightly coupled to the code anyways (lots of parameters).